### PR TITLE
feat: Custom Encoder & Decoder

### PR DESCRIPTION
**Pull request by matusklasovity**

### CODE REVIEW

File: Sources/GoodNetworking/CodableExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,66 @@
+//
+//  Codable.swift
+//  DepoSwiftExtensions
```
Comment (plajdo) [inline Sources/GoodNetworking/CodableExtensions.swift:3]: I'd fix the file headers (in all files) considering this is now a public library

File: Sources/Mockable/Mockabe.swift [hunk 2]
```diff
@@ -12,7 +12,7 @@ public protocol Mockable {
```
Comment (plajdo) [inline Sources/Mockable/Mockabe.swift:12]: this file has a typo in its name - `Mockabe`

File: Sources/GoodNetworking/CodableExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    
+    var encoder: JSONEncoder { get }
+    
+}
+
+public extension Encodable where Self: WithCustomEncoder {
+    
+    /// The `encode` method returns `Data` representation of the `Encodable` object.
+    /// - Throws: An error if encoding fails.
+    /// - Returns: `Data` representation of the `GREncodable` object.
```
[the remark this answers is not in the document]
Comment (andrej-jasso) [reply Sources/GoodNetworking/CodableExtensions.swift:22]: @matusklasovity  prosimta pozri dokumentaciu a najdi zmienky GREncodable GRDecodable v projekte a nahrad ich prosimta ;)

File: Tests/GoodNetworkingTests/DateDecodingTest.swift [hunk 4]
```diff
@@ -0,0 +1,80 @@
+//
+//  File.swift
```
[the remark this answers is not in the document]
Comment (andrej-jasso) [reply Tests/GoodNetworkingTests/DateDecodingTest.swift:2]: @matusklasovity ked vytvoris subor v Packag-i bohuzial pomenovanie tam neni automaticky z file wizardu pretoze akonahle vytvoris subor je vytvoreny a uz ho iba nasledne premenuvavas. Ale v hlavicke vzdy ostava File.swift a musis to zmenit

File: Sources/GoodNetworking/DataRequestExtensions.swift [hunk 5]
```diff
@@ -33,6 +33,39 @@ public extension DataRequest {
...
+    ///   - preprocessor:        `DataPreprocessor` which filters the `Data` before serialization. `PassthroughPreprocessor()`
+    ///                          by default.
+    ///   - decoder:             `DataDecoder` instance used to decode response `Data`. `JSONDecoder()` by default.
+    ///   - emptyResponseCodes:  `Set<Int>` of HTTP status codes for which empty responses are allowed. `[204, 205]` by
+    ///                          default.
+    ///   - emptyRequestMethods: `Set<HTTPMethod>` of `HTTPMethod`s for which empty responses are allowed, regardless of
+    ///                          status code. `[.head]` by default.
+    ///
+    /// - Returns:               The `DataResponsePublisher`.
+    func goodify<T: Decodable & WithCustomDecoder>(
```
Comment (andrej-jasso) [inline Sources/GoodNetworking/DataRequestExtensions.swift:63]: I would add decoder in all goodify functions as a parameter with default value. 

cc: @plajdo

**Review by andrej-jasso — CHANGES_REQUESTED** [review]

File: Sources/GoodNetworking/DataRequestExtensions.swift [hunk 6]
```diff
@@ -53,23 +55,24 @@ public extension DataRequest {
...
         preprocessor: DataPreprocessor = DecodableResponseSerializer<T>.defaultDataPreprocessor,
         emptyResponseCodes: Set<Int> = DecodableResponseSerializer<T>.defaultEmptyResponseCodes,
-        emptyResponseMethods: Set<HTTPMethod> = DecodableResponseSerializer<T>.defaultEmptyRequestMethods
+        emptyResponseMethods: Set<HTTPMethod> = DecodableResponseSerializer<T>.defaultEmptyRequestMethods,
+        decoder: JSONDecoder = (T.self as? WithCustomDecoder.Type)?.decoder ?? JSONDecoder()
     ) -> AnyPublisher<[T], AFError> {
         let serializer = DecodableResponseSerializer<[T]>(
             dataPreprocessor: preprocessor,
-            decoder: T.decoder,
+            decoder: JSONDecoder(),
```
Comment (plajdo) [inline Sources/GoodNetworking/DataRequestExtensions.swift:75]: is `decoder` parameter required, when the implementation seems to create a new instance of `JSONDecoder` every time anyways?

File: Sources/GoodNetworking/GREndPointManagerProtocol.swift [hunk 7]
```diff
@@ -1,5 +1,5 @@
 //
-//  GREndpointManager.swift
+//  GREndPointManagerProtocol.swift
```
Comment (plajdo) [inline Sources/GoodNetworking/GREndPointManagerProtocol.swift:2]: `Endpoint` should not be capitalized since it's just one word

**Review by plajdo — CHANGES_REQUESTED** [review]

File: Tests/GoodNetworkingTests/DateDecodingTest.swift [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+import Foundation
+import GoodNetworking
+import XCTest
+import Mockable
+
+final class DateDecodingTest: XCTestCase {
+        
+    struct IsoDateWrapper: Decodable, WithCustomDecoder {
+        
+        static var decoder: JSONDecoder {
```
Comment (andrej-jasso) [inline Tests/GoodNetworkingTests/DateDecodingTest.swift:17]: Trocha som sa nad tym zamyslal a vychadza mi ze mozno by bolo krajsie zaobalit tu strategiu.
Vies to aplikovat ako extension na WithCustomDecoder a zasa budeme mat ako predtym cez override. To mi prislo krajsie ako toto riesenie

**Review by andrej-jasso — CHANGES_REQUESTED** [review]
Skus si to este prejst ci to nevieme predsa skratit ten zapis. Toto mi pride dost zdlhavy zapis

**Review by andrej-jasso — APPROVED** [review]

**Review by plajdo — APPROVED** [review]

**Pull request merged**